DAT-6884: Adding changes to CreateProcedureGenerator

psrijan opened the pull request.

Commit:
DAT-6884: Adding changes to CreateProcedureGenerator

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateProcedureGenerator.java:
```diff
@@ -90,11 +93,33 @@ public ValidationErrors validate(CreateProcedureStatement statement, Database da
     }
 
+    public static void getSqlStatementsWithSet(List<Sql> sql, String procedureText, CreateProcedureStatement statement) {
+        String IS_SET_REGEX = "SET\\s(\\w)+\\s(ON|OFF);?";
```
It will not handle SET statement like `SET     QUOTED_IDENTIFIER  ON `
Also, in the ticket mentioned that we need to leave only `QUOTED_IDENTIFIER` and `ANSI_NULL`
And this regexp will add all SET SOME_OPTION ON/OFF and it does not meet ticket requirements
MSSQL SET statements - https://docs.microsoft.com/en-us/sql/t-sql/statements/set-statements-transact-sql?view=sql-server-ver15

psrijan replied:
It will accept without the semi colon as the ? at the end states taht ; is optional

achehrynets replied:
@psrijan , I re-read my comment and realized that the comment is confusing.
I meant that `SET\\s(\\w)+\\s(ON|OFF);?` regexp will not handle multiple whitespace after `SET` keyword and after configuration property. Also, it does not handle lowercased and lowercased+uppercased statements.
I'm leaving for you different SQL statements to test regexp, make sure that regexp can find those statements.
![Screenshot from 2021-12-20 13-17-14](https://user-images.githubusercontent.com/15807734/146759190-98070a2c-2da4-4e7f-9332-1422e58ba818.png)

I recommend you to use a third-party tool to test regexp, for example, I use https://rubular.com/, but note that it's for ruby which means that once you implemented regexp, you will need to adapt it to java regexp format which as usually means adding an additional slash (`/`) to existing slash (e.g. `SET\s(\w)+...` is ruby format and java format looks like `SET\\s(\\w)+...`)
You can also try to find java specific tool to test regexp.

psrijan replied:
done

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateProcedureGenerator.java:
```diff
@@ -90,11 +93,33 @@ public ValidationErrors validate(CreateProcedureStatement statement, Database da
         return sql.toArray(new Sql[sql.size()]);
     }
 
+    public static void getSqlStatementsWithSet(List<Sql> sql, String procedureText, CreateProcedureStatement statement) {
```
the method name starts with `get...` but it returns nothing (void return type) and it seems confusing.

Also, it's not only checks given SQL content for `SET ... ON/OFF` it also adds Sql object to SQL list and I think that it should be moved to separate method.

psrijan replied:
Sounds good. I will add it to a separate utility class with name that starts with add instead of set. One thing, the checking and adding are very interrelated and can be in the same function I feel.

achehrynets reviewed (changes requested):
left comments

Commit:
DAT-6884: Refining code

Commit:
DAT-6884: Adding Tests for MSSQLUtils

Commit:
DAT-6884: Adding Regex information

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/util/MSSQLUtil.java:
```diff
@@ -0,0 +1,67 @@
+ * MSSQL Util for generator to handle SET methods for now.
+ * Additional utility methods can be added here
+ */
+public class MSSQLUtil {
```
I would recommend this class to point that it's mssql util for SQLGenerators.
For example: SqlGeneratorMssqlUtil or something like that

psrijan replied:
Done!

psrijan replied:
Changed to SqlGeneratorMSSQLUtil.

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/util/MSSQLUtil.java:
```diff
@@ -0,0 +1,67 @@
+ * Additional utility methods can be added here
+ */
+public class MSSQLUtil {
+    public static String IS_SET_REGEX = "(?i)SET\\s+(?i)(ANSI_NULLS|QUOTED_IDENTIFIER)\\s(?i)(ON|OFF)(\\s*\\t*\\n*);?";
```
@bbrown-at-datical , what do you think about this regexp, will it handle all possible variations of `SET ANSI_NULLS/QUOTED_IDENTIFIER ... `?

achehrynets reviewed (commented):
Left comments

Commit:
DAT-6884: Review changes

Commit:
DAT-6884:Additional changes

Commit:
Merge branch 'ddb' of github.com:Datical/liquibase into DAT-6884

Commit:
DAT-6884: Removing code

achehrynets commented on liquibase-core/src/test/java/liquibase/sqlgenerator/core/util/MSSQLUtilTest.java:
```diff
@@ -0,0 +1,91 @@
+    public void addSqlStatementWithOneSet_relationFunction() {
+        List<Sql> sqlList = new ArrayList<>();
+        SqlGeneratorMSSQLUtil.addSqlStatementsToList(sqlList, data_amy, new View("abc", "bcd", "efg"));
+        System.out.println("--------------------------------");
```
Please, remove sout.

psrijan replied:
done

achehrynets commented on liquibase-core/src/test/java/liquibase/sqlgenerator/core/util/MSSQLUtilTest.java:
```diff
@@ -0,0 +1,91 @@
+        List<Sql> sqlList = new ArrayList<>();
+        SqlGeneratorMSSQLUtil.addSqlStatementsToList(sqlList, data1_quoted_ansi, new View("abc", "bcd", "efg"));
+        System.out.println("--------------------------------------");
+        System.out.println("Data1_QUOTED_ANSI" + sqlList.size());
```
Please, remove sout.

psrijan replied:
done

psrijan replied:
done

achehrynets commented on liquibase-core/src/main/java/liquibase/database/AbstractJdbcDatabase.java:
```diff
@@ -1271,7 +1271,6 @@ public boolean isSafeToRunUpdate() throws DatabaseException {
     @Override
     public void executeStatements(final Change change, final DatabaseChangeLog changeLog, final List<SqlVisitor> sqlVisitors) throws LiquibaseException {
         SqlStatement[] statements = change.generateStatements(this);
-
```
Please, remove the new line.

psrijan replied:
done. Added line back

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/util/SqlGeneratorMSSQLUtil.java:
```diff
@@ -0,0 +1,67 @@
+ * Additional utility methods can be added here
+ */
+public class SqlGeneratorMSSQLUtil {
+    public static String IS_SET_REGEX = "(?i)SET\\s+(?i)(ANSI_NULLS|QUOTED_IDENTIFIER)\\s+(?i)(ON|OFF)(\\s*\\t*\\n*);?";
```
```suggestion
    private static final String IS_SET_REGEX = "(?i)SET\\s+(?i)(ANSI_NULLS|QUOTED_IDENTIFIER)\\s+(?i)(ON|OFF)(\\s*\\t*\\n*);?";
```

psrijan replied:
done

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/util/SqlGeneratorMSSQLUtil.java:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    public static Matcher getMatcher(String sqlText) {
```
```suggestion
    private static Matcher getMatcher(String sqlText) {
```

achehrynets replied:
Can you please somehow rename this method to make it more obvious?

psrijan replied:
done

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/util/SqlGeneratorMSSQLUtil.java:
```diff
@@ -0,0 +1,67 @@
+    public static String IS_SET_REGEX = "(?i)SET\\s+(?i)(ANSI_NULLS|QUOTED_IDENTIFIER)\\s+(?i)(ON|OFF)(\\s*\\t*\\n*);?";
+
+    /**
+     * General add Sql Statement Util
```
Can you please add a better description of this method?

psrijan replied:
done

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/util/SqlGeneratorMSSQLUtil.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    /**
+     * View Specific
```
Can you please add a better description of this method?

psrijan replied:
done

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/util/SqlGeneratorMSSQLUtil.java:
```diff
@@ -0,0 +1,67 @@
+            hasSetMatcher.reset(); // reset to start to make the find and replace fit in a single loop
+            while (hasSetMatcher.find()) {
+                String curSetSql = hasSetMatcher.group();
+                sql.add(new UnparsedSql(curSetSql, ";")); // adding a default end delimiter of ;
```
Can a user specify a custom delimiter that is different from `;`?

achehrynets commented on liquibase-core/src/test/java/liquibase/sqlgenerator/core/util/MSSQLUtilTest.java:
```diff
@@ -0,0 +1,91 @@
+import java.util.List;
+
+public class MSSQLUtilTest {
+    private final static String data1 = "CREATE PROCEDURE [s1].[p10]   \n" +
```
```suggestion
    private static final String CREATE_PROCEDURE_DDL = "CREATE PROCEDURE [s1].[p10]   \n" +
```

psrijan replied:
done

achehrynets commented on liquibase-core/src/test/java/liquibase/sqlgenerator/core/util/MSSQLUtilTest.java:
```diff
@@ -0,0 +1,91 @@
+            "    SELECT '10';\n" +
+            "    SELECT '10\n";
+
+    private final static String data1_quoted_ansi = "CREATE PROCEDURE [s1].[p10]   \n" +
```
```suggestion
    private static final String CREATE_PROCEDURE_DDL_WITH_QUOTED_ANSI = "CREATE PROCEDURE [s1].[p10]   \n" +
```

achehrynets commented on liquibase-core/src/test/java/liquibase/sqlgenerator/core/util/MSSQLUtilTest.java:
```diff
@@ -0,0 +1,91 @@
+            "\t        DROP PROCEDURE \n" +
+            "\t        string';\n";
+
+    private final static String data_amy = "SET QUOTED_IDENTIFIER OFF; \n" +
```
```suggestion
    private static final String CREATE_PROCEDURE_DDL_WITH_QUOTED_ANSI_AMY_EXAMPLE = "SET QUOTED_IDENTIFIER OFF; \n" +
```

achehrynets commented on liquibase-core/src/test/java/liquibase/sqlgenerator/core/util/MSSQLUtilTest.java:
```diff
@@ -0,0 +1,91 @@
+            "SELECT \"Amy\";\n"+
+            "END;\n";
+
+    private final static String test_proc = "SET ANSI_NULLS ON\n" +
```
```suggestion
    private static final String CREATE_PROCEDURE_DDL_WITH_USER_SETTINGS = "SET ANSI_NULLS ON\n" +
```

psrijan replied:
done

achehrynets commented on liquibase-core/src/test/java/liquibase/sqlgenerator/core/util/MSSQLUtilTest.java:
```diff
@@ -0,0 +1,91 @@
+
+    private final static String data1_quoted_ansi = "CREATE PROCEDURE [s1].[p10]   \n" +
+            "AS   \n" +
+            "    SET QUOTED_IDENTIFIER ON;    \n" +
```
This does not seem like valid sql, not sure that `SET QUOTED_IDENTIFIER ON` and `SET ANSI_NULLS ON` can be in the procedure body.
Can you run this SQL manually to verify it?

psrijan replied:
removed

achehrynets commented on liquibase-core/src/test/java/liquibase/sqlgenerator/core/util/MSSQLUtilTest.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Test
+    public void addSqlStatementWithOneSet_testProc() {
```
Please, use camel case because it's java standard.

psrijan replied:
done

achehrynets commented on liquibase-core/src/test/java/liquibase/sqlgenerator/core/util/MSSQLUtilTest.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Test
+    public void addSqlStatementWithOneSet_relationFunction() {
```
Please, use camel case because it's java standard.

psrijan replied:
Done

achehrynets commented on liquibase-core/src/test/java/liquibase/sqlgenerator/core/util/MSSQLUtilTest.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Test
+    public void addSqlStatementsWithTwoSet_relationFunction() {
```
Please, use camel case because it's java standard.

psrijan replied:
Done

achehrynets reviewed (changes requested):
@psrijan, please review my comments.

andrulban commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/util/SqlGeneratorMSSQLUtil.java:
```diff
@@ -0,0 +1,67 @@
+     * @param sqlText
+     * @param delimiter
+     */
+    public static void addSqlStatementsToList(List<Sql> sql, String sqlText, String delimiter) {
```
Looks like we can make this method simpler, for me looks like code below will do the same, but with less if/else statements:

```
   public static void addSqlStatementsToList(List<Sql> sql, String sqlText, String delimiter) {
        Matcher hasSetMatcher =  getMatcher(sqlText);
        String cleanSqlText = sqlText;
        while (hasSetMatcher.find()) {
                String curSetSql = hasSetMatcher.group();
                sql.add(new UnparsedSql(curSetSql, delimiter));
                cleanSqlText = cleanSqlText.replace(curSetSql, "");
        }
        sql.add(new UnparsedSql(cleanSqlText, delimiter));
    }
```

psrijan replied:
Yea. I think what you said makes sense here.

psrijan replied:
done

andrulban commented on liquibase-core/src/test/java/liquibase/sqlgenerator/core/util/MSSQLUtilTest.java:
```diff
@@ -0,0 +1,91 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class MSSQLUtilTest {
```
Could you please add a few test cases?
1. Where SQL is not upper case, so case sensitivity is also checked
2. Where SET statement is in the same line as usual statement (e.g. INSERT)

psrijan replied:
done

Commit:
DAT-6884: Additional test cases added, renamed functions, better descriptions

Commit:
DAT-6884: Adding changes to regex logic

Commit:
DAT-6884: Additional changes

andrulban committed:
[DAT-6884] Rewrite addSqlStatementsToList algorithm

Fixed bugs:
1. [Wrong order] Set statements were always first, which causes issues when we have both on and off in the same script
2. [Replacing multiple statements] When we have more than one similar set statement it was replaced after first look up as a result - missing set statements
3. [Empty string in result list] If there were blank lines on the beginning or on the end whereas there are SET statements only - then script would end up with empty string line in the list

andrulban reviewed (commented):
Hi @achehrynets, could you please take a look? I have rewritten algorithm to fix 3 bugs (all described in the commit description) which I have found (one of them was mentioned in the ticket by @TymofiiKritsak ). I will push ticket back to test, but will make sure that got approval from you on new algorithm.

andrulban committed:
[DAT-6884] Rewrite set statements split logic with StringClauses.ClauseIterator instead of regex

Now it goes word by word and splits set statements and remove them from original string including delimiters and later places them in correct order into SQL list

andrulban committed:
[DAT-6884] MssqlSplitStatements into a separate class

Commit:
Merge branch 'ddb' into DAT-6884

Commit:
Merge branch 'DAT-6884' of github.com:Datical/liquibase into DAT-6884

andrulban committed:
[DAT-6884] Move MssqlSplitStatements back to CreateProcedureGenerator

achehrynets reviewed (approved):
@andrulban , changes look great.
May I ask you to add java docs for the new class and methods?

andrulban committed:
[DAT-6884] Add JavaDoc comments

andrulban committed:
[DAT-6884] Rewrite set statements split logic to split out only statements before and do not split after

andrulban committed:
[DAT-6884] Add "SET ANSI_NULLS ON" and "SET QUOTED_IDENTIFIER ON" before executing any CreateProcedure and Create View generator to keep default values for the setting and isolate scripts behaviours

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateProcedureGenerator.java:
```diff
@@ -19,9 +19,13 @@
 
 public class CreateProcedureGenerator extends AbstractSqlGenerator<CreateProcedureStatement> {
+
+    private static final List<String> BEFORE_BODY_START_STATEMENTS = Arrays.asList("CREATE", "ALTER");
```
We use the constant only in one place, what about removing it?

andrulban replied:
Done

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateProcedureGenerator.java:
```diff
@@ -167,4 +189,109 @@ public static String addSchemaToText(String procedureText, String schemaName, St
+
+    private static Object splitOutIfSetStatement(Object next, StringClauses.ClauseIterator clauseIterator,
+                                                 String endDelimiter, ArrayList<String> setStatements) {
+        while (next != null && next.toString().equalsIgnoreCase("SET")) {
```
```suggestion
        while ("SET".equalsIgnoreCase(next)) {
```

andrulban replied:
"SET".equalsIgnoreCase(next) <- It does not compile, because next is an object
"SET".equalsIgnoreCase((String) next) <- it will compile but we don't have guarantee that next is String, so we might get an error
"SET".equalsIgnoreCase(next.toString()) <- will work, but if it is null, then it will throw NPE, so current check looks optimal

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateProcedureGenerator.java:
```diff
@@ -167,4 +189,109 @@ public static String addSchemaToText(String procedureText, String schemaName, St
+            clauseIterator.remove();
+            next = clauseIterator.nextNonWhitespace();
+
+            if (next != null
+                    && (next.toString().equalsIgnoreCase("ANSI_NULLS")
+                    || next.toString().equalsIgnoreCase("QUOTED_IDENTIFIER"))) {
```
```suggestion
            if ("ANSI_NULLS".equalsIgnoreCase(next) || "QUOTED_IDENTIFIER".equalsIgnoreCase(next)) {
```

andrulban replied:
Same case as above

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateProcedureGenerator.java:
```diff
@@ -167,4 +189,109 @@ public static String addSchemaToText(String procedureText, String schemaName, St
+                clauseIterator.remove();
+                next = clauseIterator.nextNonWhitespace();
+
+                if (next != null
+                        && (next.toString().equalsIgnoreCase("ON")
+                        || next.toString().equalsIgnoreCase("OFF"))) {
```
```suggestion
                if ("ON".equalsIgnoreCase(next) || "OFF".equalsIgnoreCase(next)) {
```

andrulban replied:
Same case as above

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateProcedureGenerator.java:
```diff
@@ -167,4 +189,109 @@ public static String addSchemaToText(String procedureText, String schemaName, St
+                }
+            }
+
+            if (!bufferedSetStatement.toString().isEmpty() && !bufferIsUsed) {
```
```suggestion
            if (StringUtils.isNotEmpty(bufferedSetStatement) && !bufferIsUsed) {
```

andrulban replied:
StringUtils.isNotEmpty(bufferedSetStatement.toString()) <- done

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateViewGenerator.java:
```diff
@@ -93,7 +105,15 @@ public ValidationErrors validate(CreateViewStatement createViewStatement, Databa
             }
         }
+
+        if (mssqlSetStatementsBefore != null && !mssqlSetStatementsBefore.isEmpty()) {
```
```suggestion
        if (CollectionUtils.isNotEmpty(mssqlSetStatementsBefore)) {
```

achehrynets replied:
Will work in case if liquibase has apache commons collections dependency.

andrulban replied:
Unfortunately Liquibase does not have this dependency :(

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateViewGenerator.java:
```diff
@@ -48,14 +51,23 @@ public ValidationErrors validate(CreateViewStatement createViewStatement, Databa
     public Sql[] generateSql(CreateViewStatement statement, Database database, SqlGeneratorChain sqlGeneratorChain) {
+        List<Sql> sql = new ArrayList<Sql>();
+        List<String> mssqlSetStatementsBefore = new ArrayList<>();
 
```
Can you add property to set default delimiter and use it in this part of code?

andrulban replied:
Done

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateViewGenerator.java:
```diff
@@ -48,14 +51,23 @@ public ValidationErrors validate(CreateViewStatement createViewStatement, Databa
 
     @Override
     public Sql[] generateSql(CreateViewStatement statement, Database database, SqlGeneratorChain sqlGeneratorChain) {
+        List<Sql> sql = new ArrayList<Sql>();
+        List<String> mssqlSetStatementsBefore = new ArrayList<>();
```
```suggestion
        String defaultDelimiter = ";";
        List<Sql> sql = new ArrayList<Sql>();
        List<String> mssqlSetStatementsBefore = new ArrayList<>();
```

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateViewGenerator.java:
```diff
@@ -48,14 +51,23 @@ public ValidationErrors validate(CreateViewStatement createViewStatement, Databa
+        String body = statement.getSelectQuery();
+        if (database instanceof MSSQLDatabase) {
+            CreateProcedureGenerator.MssqlSplitStatements mssqlSplitStatements =
+                    splitSetStatementsOutForMssql(body, ";", Collections.singletonList("VIEW"));
```
```suggestion
                    splitSetStatementsOutForMssql(body, defaultDelmiter, Collections.singletonList("VIEW"));
```

achehrynets commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateViewGenerator.java:
```diff
@@ -93,7 +105,15 @@ public ValidationErrors validate(CreateViewStatement createViewStatement, Databa
+        if (mssqlSetStatementsBefore != null && !mssqlSetStatementsBefore.isEmpty()) {
+            mssqlSetStatementsBefore
+                    .forEach(mssqlSetStatement ->
+                            sql.add(new UnparsedSql(mssqlSetStatement, ";")));
```
```suggestion
                            sql.add(new UnparsedSql(mssqlSetStatement, defaultDelmiter)));
```

achehrynets reviewed (changes requested):
@andrulban , left comments

andrulban committed:
[DAT-6884] Remove constant

andrulban committed:
[DAT-6884] Move "SET ANSI_NULLS ON" and "SET QUOTED_IDENTIFIER ON" from before statements to the end statements

It will fix issue with such scenario:
1. Generate changelog from existing DB where quoted identifiers and nulls were off <- it will add <set> OFF changes into changeset before executing main chain out of the changeset
2. Drop DB and update DB from this changelog <- before this change it will switch ON identifiers before execution for isolation, but now it will switch them ON after execution and it wil keep isolation and won't influence out current logic
3. Generate changelog <- with previous logic it won't contain <set> changes in changeset, but with new changes will contain it
